### PR TITLE
Check if email_verified

### PR DESCRIPTION
**Pull request by donaldgray**
Reject requests from users who have not had their email verified.

### CODE REVIEW

File: src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt [hunk 1]
```diff
@@ -7,6 +7,8 @@ import org.springframework.http.converter.HttpMessageNotReadableException
 import org.springframework.security.core.annotation.AuthenticationPrincipal
 import org.springframework.web.bind.annotation.*
 import uk.bfi.uvaudit.security.AuditUser
+import javax.naming.AuthenticationException
```
Comment (garyttierney) [inline src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt]: I think you want [`org.springframework.security.core.AuthenticationException`](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/core/AuthenticationException.htmll) here, `javax.naming.AuthenticationException`refers to authentication errors with JNDI or directory services.

**Review by garyttierney — CHANGES_REQUESTED** [review]
Just one small comment.

File: src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt [hunk 2]
```diff
@@ -26,9 +28,16 @@ class AuditEventController(
         logger.error("Audit event structure was invalid", ex)
     }
 
+    @ExceptionHandler(AuthenticationException::class)
+    @ResponseStatus(HttpStatus.UNAUTHORIZED)
```
Comment (danielgrant) [inline src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt]: I would suggest this is changed to a `403 Forbidden` instead. The `401 Unauthorized` is used to indicate that the user is not authenticated. In this case, the user is authenticated, but we're denying them access all the same.

https://auth0.com/blog/forbidden-unauthorized-http-status-codes/

File: src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt [hunk 3]
```diff
@@ -39,6 +48,15 @@ class AuditEventController(
...
         @RequestHeader("X-API-RequestURI") requestUri: String
     ) {
+        validateUser(user)
         writer.write(user.id, requestType, requestUri)
     }
+
+    private fun validateUser(
+        user: AuditUser
+    ) {
+        if (user.getAttribute<Boolean>("email_verified") != true) {
```
Comment (danielgrant) [inline src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt:58]: Is there a subtlety I'm missing that means this won't work?

```
if (!user.getAttribute<Boolean>("email_verified"))
```

**Review by danielgrant — CHANGES_REQUESTED** [review]

File: src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt (hunk 3 above)
Comment (garyttierney) [reply src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt:58]: `getAttribute<T>` returns a `T?` (i.e. nullable), so you'd need a non-null assertion on the result of the call first to write it like that.

File: src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt (hunk 1 above)
Comment (donaldgray) [reply src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt]: `org.springframework.security.core.AuthenticationException` is abstract so I added a custom exception type `EmailNotVerifiedException` which inherits from it

File: src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt (hunk 2 above)
Comment (donaldgray) [reply src/main/kotlin/uk/bfi/uvaudit/event/AuditEventController.kt]: Done

**Review by danielgrant — APPROVED** [review]

**Review by danielgrant — APPROVED** [review]

**Pull request merged**